service/lib/dinstaller/dbus/service_runner.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2022] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "eventmachine"
+require "dinstaller/manager"
+require "dinstaller/dbus/service"
+
+module DInstaller
+  module DBus
+    # Set up and run a given D-Bus service
+    #
+    # @example Run the manager service
+    #   runner = ServiceRunner.new(:manager)
+    #   runner.run
+    #
+    # @example Run the users service
+    #   runner = ServiceRunner.new(:users)
+    #   runner.run
+    class ServiceRunner
+      # @param name [Symbol,String] Service name (:manager, :users, etc.)
+      # @param logger [Logger] Service logger
+      def initialize(name, logger: Logger.new($stdout))
+        @name = name || :manager
+        @logger = logger
+      end
+
+      # Run the Service
+      #
+      # This method listens for D-Bus calls.
+      def run
+        service = if name == :manager
+          setup_manager(logger)
+        else
+          setup_service(name, logger)
+        end
+
+        service.export
+        EventMachine.run do
+          EventMachine::PeriodicTimer.new(0.1) { service.dispatch }
+        end
+      end
+
+    private
+
+      attr_reader :name, :logger
+
+      # Set up the manager service
+      #
+      # @param logger [Logger] Service logger
+      # @return [DInstaller::DBus::Service] Manager service
+      def setup_manager(logger)
+        manager = DInstaller::Manager.new(logger)
+        manager.setup
+        service = DInstaller::DBus::Service.new(manager, logger)
+        manager.probe # probe after export, so we can report on DBus
+        manager.progress.on_change { service.dispatch } # make single thread more responsive
+        service
+      end
+
+      # Set up a service
+      #
+      # @param name [String] Service name (ie, "users")
+      # @param logger [Logger] Service logger
+      # @return [#export,#dispatch] Class that implements #export and #dispatch methods.
+      def setup_service(name, logger)
+        require "dinstaller/dbus/#{name}_service"
+        klass = DInstaller::DBus.const_get("#{name.capitalize}Service")
+        klass.new(logger)
```

Comment:
Maybe we could move the logic of `setup_manager` to the constructor of `Dinstaller::DBus::Service` and initialize all the services in the very same way.

Reply:
To be honest, I do not have a clear view of how the `Manager` class will evolve now that we are moving some pieces outside. IMHO `Manager` does too much (sets up Cockpit, initializes other services, etc.).

I see your point, and perhaps it is the right approach. But by now, I prefer to inject the already initialized manager into the `Service`, so the `Service` has nothing to do (just to serve as an API on top). This piece of logic lives in a kind of ancillary class (`ServiceRunner`) and it should be pretty straightforward to move it later.

But I might be wrong.

Reply:
The thing I do not feel comfortable with is that, in other cases, the services are taking care of initializing the backends. And I do not know if I like that.